sydent/http/httpclient.py

```diff
@@ -79,115 +68,120 @@ def post_json_get_nothing(self, uri, post_json, opts):
         )
         defer.returnValue(response)
 
-class SRVClientEndpoint(object):
-    def __init__(self, reactor, service, domain, protocol="tcp",
-                 default_port=None, endpoint=HostnameEndpoint,
-                 endpoint_kw_args={}):
-        self.reactor = reactor
-        self.domain = domain
+class SimpleHttpClient(HTTPClient):
+    """
+    A simple, no-frills HTTP client based on the class of the same name
+    from Synapse.
+    """
+    def __init__(self, sydent, endpoint_factory=None):
+        self.sydent = sydent
+        if endpoint_factory is None:
+            # The default endpoint factory in Twisted 14.0.0 (which we require) uses the
+            # BrowserLikePolicyForHTTPS context factory which will do regular cert validation
+            # 'like a browser'
+            self.agent = Agent(
+                reactor,
+                connectTimeout=15,
+            )
+        else:
+            self.agent = Agent.usingEndpointFactory(
+                reactor,
+                endpoint_factory,
+            )
 
-        self.endpoint = endpoint
-        self.endpoint_kw_args = endpoint_kw_args
+class FederationContextFactory(object):
+    def getContext(self):
+        context = SSL.Context(SSL.SSLv23_METHOD)
+        try:
+            _ecCurve = crypto.get_elliptic_curve(_defaultCurveName)
+            context.set_tmp_ecdh(_ecCurve)
+        except Exception:
+            logger.exception("Failed to enable elliptic curve for TLS")
+        context.set_options(SSL.OP_NO_SSLv2 | SSL.OP_NO_SSLv3)
 
-    @defer.inlineCallbacks
-    def lookup_server(self):
-        service_name = "%s.%s.%s" % ('_matrix', '_tcp', self.domain)
+        context.set_cipher_list("!ADH:HIGH+kEDH:!AECDH:HIGH+kEECDH")
+        return context
+
+class FederationHttpClient(HTTPClient):
+    def __init__(self, sydent):
+        self.sydent = sydent
+        self.agent = MatrixFederationAgent(
+            reactor,
+            ClientTLSOptionsFactory(),
+        )
 
-        default = self.domain, 8448
+def _tolerateErrors(wrapped):
```

Comment:
I'd suggest sticking all this gubbins to do with ClientTLSOptionsFactory in a separate file